zenoh/src/api/bytes.rs

```diff
@@ -86,48 +79,59 @@ impl From<OptionZBytes> for Option<ZBytes> {
     }
 }
 
-/// Trait to encode a type `T` into a [`ZBytes`].
-pub trait Serialize<T> {
-    type Output;
+pub trait SerializedSize {
+    const FIXED_SIZE: Option<usize> = None;
+}
+impl<T: SerializedSize> SerializedSize for &T {
+    const FIXED_SIZE: Option<usize> = T::FIXED_SIZE;
+}
+
+pub enum Serialized<'a> {
+    Slice(&'a [u8]),
+    ZBytes(ZBytes),
+}
+
+impl<'a> From<&'a [u8]> for Serialized<'a> {
+    fn from(value: &'a [u8]) -> Self {
+        Self::Slice(value)
+    }
+}
 
-    /// The implementer should take care of serializing the type `T` into a [`ZBytes`].
-    /// If [`Encoding`] metadata is important in a given system, the caller should take care
-    /// of setting the proprer [`Encoding`] value when calling functions like [`crate::Session::put`].
-    fn serialize(self, t: T) -> Self::Output;
+impl From<ZBytes> for Serialized<'_> {
+    fn from(value: ZBytes) -> Self {
+        Self::ZBytes(value)
+    }
 }
 
-pub trait Deserialize<T> {
-    type Input<'a>;
+pub trait Serialize<'a>: Sized + SerializedSize {
     type Error;
+    fn write(self, writer: &mut ZBytesWriter) -> Result<(), Self::Error>;
+    fn serialize(self) -> Result<Serialized<'a>, Self::Error>;
+}
 
-    /// The implementer should take care of deserializing a [`ZBytes`] into the type `T`.
-    /// If [`Encoding`] metadata is required in a given system, the caller should take care of checking the proprer
-    /// [`Encoding`] value (e.g. [`crate::sample::Sample::encoding`]) to select the right type `T` to deserialize into.
-    fn deserialize(self, t: Self::Input<'_>) -> Result<T, Self::Error>;
+pub trait Deserialize<'a>: Sized + SerializedSize {
+    fn read(reader: &mut ZBytesReader) -> Result<Self, ZDeserializeError>;
+    fn deserialize(zbytes: &'a ZBytes) -> Result<Self, ZDeserializeError>;
+    fn try_from_zbytes(zbytes: ZBytes) -> Result<Self, ZDeserializeError>;
```

Comment:
What's the purpose of `try_from_zbytes` as part of the trait? It seems not needed to me.

Reply:
Even if everything will be dropped so we don't care anymore, here was the reason:
`deserialize` couldn't be call in the context of an iterator when you deserialize first a `ZBytes`, which would then have an anonymous lifetime. So deserializing an iterator of `Cow<'a, [u8]>` was not possible as `'a` would be anonymous in the implementation.
That's why two different methods were needed, one for `ZBytes::deserialize`, and the other one for the iterator.